designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+
+#### Overriding Configuration Based on File Patterns
+
+Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `.eslint.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:
+
+```yaml
+plugins: ["react"]
+rules:
+   react/jsx-uses-react: error
+    semi: error
+
+overrides:
+    - files: "*.md"
+      plugins: ["markdown"],
+      processor: "markdown/markdown"
+```
+
+This can be written in `.eslint.js` as an array of two configs:
+
+```js
+exports.config = [
+    {
+        files: "*.js",
+        ruledefs: {
+            react: require("eslint-plugin-react").rules,
+        },
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
+```
+
+When ESLint uses this config, it will check each `files` pattern to determine which config applies. In this way, returning an array acts exactly the same as the array in `overrides`, with the exception that no configuration options are inherited.
+
+#### Replacing `.eslintignore`
+
+Because there is only one `.eslint.js` file to consider, ESLint doesn't have to first search directories to determine its location. That allows `.eslint.js` to specify files to ignore directly instead of relying on `.eslintignore`. For backwards compatibility, users could create a config like this:
+
+```js
+const fs = require("fs");
+
+exports.config = {
+    files: "*.js",
+    ignore: fs.readFileSync(".eslintignore").replace(/\n/g, ";")
+};
+```
+
+### The `@eslint/config` Package
+
+Because some of the operations ESLint currently do are quite complicated, this design includes a utility package called `@eslint/config` that users can use for common tasks. This package contains the following utilities:
+
+1. `Config` - a class that both wraps a configuration object to provide helper methods for manipulating configs and a container for static methods to work with configs.
+1. `RuleLoader` - a class that aids in loading rules from a directory.
+
+The `@eslint/config` package is intended to add back functionality that would be removed from ESLint with this design.
+
+**Note:** It is not required to use this package with `.eslint.js`; this is an optional tool that users may choose to use for convenience.
+
+#### `Config` Class
+
+The `Config` class has this form:
+
+```js
+class Config {
+
+    // pass in any object literal to wrap in Config
+    constructor(object) {}
+
+    // convenience method for easily setting rule severity
+    setRuleSeverity(ruleId, severity) {}
+
+    // determine if an object is a Config
+    static isConfig(object) {}
+
+    // create a new config based on another config
+    static create(baseConfig, overrideConfig) {}
```

Comment:
Can I use an array as the `baseConfig`?

Currently, we can use `overrides` property in shared configuration. I'm a heavy user of that. 

```yml
root: true
extends:
  - plugin:@mysticatea/es2018
```

The above config has override settings for each `*.vue`, `*.ts`, `*.tsx`, `test/**`, `scripts/**`, `rollup.config.js`, and `webpack.config.js`.

Since new configuration system replaces the `overrides` property to an array of configs, it will be much if the `baseConfig` doesn't accept an array.

Reply:
...how about using an array instead of `assign`?

```js
const { configs } = require("@mysticatea/eslint-plugin")

exports.config = [
    // corresponding `extends: "plugin:@mysticatea/es2018"`
    ...configs.es2018,

    // corresponding `overrides`
    {
        files: "*.js",
        rules: {
            semi: ["error", "always"]
        }
    }
]
```

Because the current `overrides` merges all matched entries, I'm imaging the array config in the new system is same.
In this case, we can use it as substitude of `extends` property. Also, the string entry for embedded presets will be useful.

```js
exports.config = [
    "eslint:recommended",
    {
        files: "*.js",
        rules: {
            semi: "error"
        }
    }
]
```

Reply:
So the solution I'm imagining is:

```js
// Merge all matched entries.
const actualConfig = Config.create(...configs.filter(config => match(filePath, config)))
```

This is a very simple rule: the setting at the last entry is prior to others.
I don't think that it's difficult to understand, as different from the current `overrides` property and `extends` property.

Reply:
I see, thanks for explaining. Can you take a look at the latest proposal that adds back `extends` and see if that addresses your concerns?

Reply:
Sorry, I just read this over again and I think I'm not understanding correctly. I think what you're saying is:

In the current system, if a shareable config has `overrides` then  those `overrides` are merged into a single list when `extends` is used, and if two `files` patterns match, then those overrides are merged together? Is that correct?

And in this proposal, because there are no overrides, you lose that ability?

If I'm understanding those points correctly, then I agree, that's a problem we need to solve. I'm not quite sure what you're referring to in that code example, though. Could you write out a more complete example?

Reply:
> Can you take a look at the latest proposal that adds back extends and see if that addresses your concerns?

Sure. I guess that the `extends` property is not going to work properly. How does it work on the following case?

```js
exports.config = {
    extends: [
        // A shareable config which has overriding (= an array).
        [
            { files: "*.js", rules: { semi: "error" } },
            { files: "*.ts", parser: require("@typescript-eslint/parser")}
        ]
    ],
    files: "*.js"
    rules: { semi: ["error", "never"] }
}
```

Where does the extended TS config go?
How do I overwrite the TS config?

In the new system, every config entry has `files` property. I don't think that `extends` property with `overrids` functionality makes sense.

----

> In the current system, if a shareable config has overrides then those overrides are merged into a single list when extends is used, and if two files patterns match, then those overrides are merged together? Is that correct?
>
> And in this proposal, because there are no overrides, you lose that ability?

Yes, you are right.

----

> Could you write out a more complete example?

Sure. My code was the code to determine the actual config in ESLint core.

```js
/** Load eslint.config.js */
function loadConfig(cwd) {
    const configOrConfigs = require(path.resolve(cwd, "eslint.config"))
    const configs = Array.isArray(configOrConfigs) ? configOrConfigs : [configOrConfigs]
    // TBD: Validation
    return configs
}

/** Get actual config for a given file. */
function getActualConfig(configs, filePath) {
    return Config.assign({}, ...configs.filter(c => match(c, filePath)))
}
```

There is an assumption: both `extends` and `overrides` properties do not exist.

For each linting target file, ESLint merges all matched config entries in linear. This simple behavior satisfies all of `extends` functionality, `overrides` functionality, and overwriting for each project.

For example:

```js
exports.config = [
    // A shareable config which has overriding (= an array).
    ...[
        { files: "*.js", rules: { semi: "error" } },
        { files: "*.ts", parser: require("@typescript-eslint/parser")}
    ],
    // For this project.
    {
        files: "*.js"
        rules: { /* ... */ }
    },
    {
        files: "*.ts"
        rules: { /* ... */ }
    },
]
```

And the examples of https://github.com/eslint/rfcs/pull/9#discussion_r250039318 are valid.

Reply:
Ah ok, I see what you mean. Thanks for explaining. Let me think about this for a bit and I'll respond back here when I've fully thought this through.

Reply:
Okay, after getting some good sleep, I think I understand this fully now. Your suggestion makes a lot of sense and I'm working on getting it into this proposal.